Support for np.fill_diagonal

rjenc29 opened the pull request:
<!--

Thanks for wanting to contribute to Numba :)

First, if you need some help or want to chat to the core developers, please
visit https://gitter.im/numba/numba for real time chat or post to the Numba
mailing list https://groups.google.com/a/continuum.io/forum/#!forum/numba-users.

Here's some guidelines to help the review process go smoothly.

0. Please write a description in this text box of the changes that are being
   made.

1. Please ensure that you have written units tests for the changes made/features
   added.

2. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

3. If your pull request is not ready for review but you want to make use of the
   continuous integration testing facilities please label it with `[WIP]` and
   then remove the label when you'd like it to be reviewed.

4. Once review has taken place please do not add features or make changes out of
   the scope of those requested by the reviewer (doing this just add delays as
   already reviewed code ends up having to be re-reviewed/it is hard to tell
   what is new etc!). Further, please do not rebase your branch on master/force
   push/rewrite history, doing any of these causes the context of any comments
   made by reviewers to be lost. If conflicts occur against master they should
   be resolved by merging master into the branch used for making the pull
   request.

Many thanks in advance for your cooperation!

-->
Support for np.fill_diagonal.

rjenc29 commented:
Initial commit to see if CI flags any gremlins.  Plenty left to do.

rjenc29 commented:
**Summary of progress**
---

1. **Seg faults on CI**
I have not experienced a seg fault running locally on any of my machines (ranging from quite modern to quite decrepit) - however, I have not yet had a clean CI as at least two (sometimes more) builds fail.  
The algorithm I am using is currently pretty simple and I am not using any 'close the metal' tricks.  I found that if I break down the tests to the point where each executes a small number (or one) test case (i.e. one set of inputs in the call to the jitted function) then everything passes.  As soon as I try to iterate through many permutations, at least one test seg faults on at least one CI build.  

2. **Unsafe type casting**
If you define an input array as an np.int32 (for example) and a fill diagonal value of np.nan (for example) then numpy raises, which seems reasonable.  I am not sure how to replicate this: if I naively set some coordinate in the input array directly, I find the resulting value is just set to the minimum permitted value for the datatype (I have a skipped test showing this).  If I were in python-land, I would be tempted to do something like the following:

```python
val = arr.dtype.type(val)
```

3. **numpy permits not scalar fill value**
The docs describe the fill value as `val : scalar - Value to be written on the diagonal, its type must be compatible with that of the array a.`  This scalar part is not in fact true and you can pass in a sequence / array and the array can be any shape you like.  I have replicated this behaviour (I think), but will work through some more edge cases.  

4. **bool(wrap)**
Numpy accepts anything as a value for the kwarg `wrap` - or rather, anything bool-able.  I have implemented wrap, but it requires a bool.

5. **Occasional error in logs**
About 1 in 5 times, I see this in the logs after running the tests I added locally (all of which pass):

```
Exception ignored in: <bound method ObjectRef.__del__ of <llvmlite.binding.executionengine.ExecutionEngine object at 0x0000024D74263E48>>
Traceback (most recent call last):
  File "C:\Users\rjenc\Anaconda3\envs\numbaenv\lib\site-packages\llvmlite\binding\ffi.py", line 243, in __del__
  File "C:\Users\rjenc\Anaconda3\envs\numbaenv\lib\site-packages\llvmlite\binding\ffi.py", line 204, in close
  File "C:\Users\rjenc\Anaconda3\envs\numbaenv\lib\site-packages\llvmlite\binding\executionengine.py", line 205, in _dispose
  File "C:\Users\rjenc\Anaconda3\envs\numbaenv\lib\site-packages\llvmlite\binding\ffi.py", line 107, in __call__
OSError: exception: access violation reading 0xFFFFFFFFFFFFFFFF
```

rjenc29 commented:
Out of curiosity, I split up the tests a little and the last CI was clean (further to point 1 above).  So it could be that the way I had originally framed the tests doesn't really support spamming with permutations of input types (or else isn't guaranteed to behave).

stuartarchibald commented:
Hi @rjenc29,

Thanks for working on the implementation of this feature. It seems like you've got the code working ok now?

1. That there are segfaults is worrying, I'm going to take a look, chances are that it's not something explicitly you've done though, unless there's some e.g. invalid access causing corruption.

2. The feature you want for 2. `val = arr.dtype.type(val)` is in the pending https://github.com/numba/numba/pull/3088. You can also `raise` from within the code gen section of the overload, for example:

```python
from numba.extending import overload
from numba import njit
import numpy as np

def foo(a):
  pass

@overload(foo)
def foo_impl(a):
    def impl(a):
        if np.isnan(a):
            raise ValueError("'a' is bad, this must be a str const!")
        else:
            return a + 1
    return impl

@njit
def use_foo(a):
    return foo(a)

print(use_foo(10))
print(use_foo(np.nan))
```
which gives:
```
$ python issue3145.py                                                                                             
11                                                                                                                              
Traceback (most recent call last):                                                                                              
  File "issue3145.py", line 22, in <module>                                                                                     
    print(use_foo(np.nan))                                                                                                      
ValueError: a is bad, this must be a str const! 
```
3. Yes, please try and replicate NumPy or if it's just too onerous to get it to work, perhaps cover the common cases and then we'll take a view on the rest, might be cheaper to simply raise as unsupported for now. Seems like the erroneous documentation problem is already noted in issue https://github.com/numpy/numpy/issues/6767. If you are happy with what it ought to be, perhaps raise a PR for NumPy to fix?

4. That's great, thanks. I'll take a look, we can probably just document this and work on it more later.

5. Is likely related to 1. seems like there's some corruption?

stuartarchibald commented:
Think this line is the cause of the issue:
https://github.com/numba/numba/pull/3145/files#diff-4b9c6924f184d92f076cc50b84bcc442R949

specifically the store to the `.flat` index. I would guess it is OOB *or* there is something wrong with the `.flat` impl or assignment to the same, slightly suspicious it is the latter.

stuartarchibald commented:
I remain suspicious of `.flat`, however this test (at least, because I know there's another in `test_fill_diagonal_oddballs_non_scalar_val_3`)  has OOB access (writes to index 25 of a possible 15) https://github.com/numba/numba/pull/3145/files#diff-ccc7b7b3e4a2e4c6617204d29aae45e7R585

Perhaps add in:
```
    assert (end <= len(a.flat))
```
just ahead of this loop to debug?
https://github.com/numba/numba/pull/3145/files#diff-4b9c6924f184d92f076cc50b84bcc442R948

rjenc29 commented:
Thanks for your input on this - good suggestions and I made a little tweak - let's see now CI gets on.

rjenc29 commented:
Status update
---
1. Fixed, I reckon.  An early commit I made introduced a possibility for OOB access violation (depending on the shape of the inputs), which I overlooked and have now fixed.  I have had no issues locally or on CI since and I have added a bunch more tests / combinations of input / deliberately weird inputs.

2. I have added some basic handling for unsafe type coercions; the implementation is a bit spidery, but it can be cleaned up after #3088.  I've added tests to illustrate the types of thing which get caught - it's not 100% consistent with Numpy, which permits some weird stuff:


```python
arr = np.zeros((3, 3), dtype=np.int16)
val_max = np.iinfo(arr.dtype).max + 10
val = np.array([1, 2, val_max], dtype=np.int64)
np.fill_diagonal(arr, val)
```

```
array([[     1,      0,      0],
       [     0,      2,      0],
       [     0,      0, -32759]], dtype=int16)
```
Let me know what you reckon.

3. Numpy behaviour replicated, as far as I can see.

4. Ditto.

5. I still see this from time to time on running tests, but it's not specific to the tests I've added.

stuartarchibald commented on numba/tests/test_array_reductions.py:
```diff
@@ -435,6 +438,112 @@ def test_nanmedian_basic(self):
+
+        def _multi_dimensional_array_variations(n):
+            for shape in _shape_variations(n):
+                yield np.zeros(shape, dtype=np.float64)
```
Might be good to add in something like:
```python
        def _multi_dimensional_array_variations_strided(n):
            for shape in _shape_variations(n):
                tmp = np.zeros(tuple([x * 2 for x in shape]), dtype=np.float64)
                slicer = [slice(0, x * 2, 2) for x in shape]
                yield tmp[slicer]
```
so as to get some slices, same goes for `val`. Might also be worth oscillating layout between 'C' and 'F' in the contig case?

rjenc29 replied:
Yep, agreed - tests added

stuartarchibald commented on numba/targets/arraymath.py:
```diff
@@ -914,6 +914,142 @@ def nanmedian_impl(arry):
+        elif isinstance(a.dtype, types.Float):
+            return fill_diagonal_impl_scalar_val_a_float
+        else:
+            return fill_diagonal_impl_scalar_val
```
Don't think the tests hit this line https://codecov.io/gh/numba/numba/compare/9fc637e3391e023b839afe00e6c86063c8cb82ad...c705734096b550f4ead5f050ca771ee884f8869f/src/numba/targets/arraymath.py#L1035 ?

rjenc29 replied:
Added tests - this should now be covered - will keep an eye out

stuartarchibald commented on numba/targets/arraymath.py:
```diff
@@ -914,6 +914,142 @@ def nanmedian_impl(arry):
+        elif isinstance(a.dtype, types.Float):
+            return fill_diagonal_impl_seq_val_a_float
+        else:
+            return fill_diagonal_impl_seq_val
```
Don't think the tests hit this line https://codecov.io/gh/numba/numba/compare/9fc637e3391e023b839afe00e6c86063c8cb82ad...c705734096b550f4ead5f050ca771ee884f8869f/src/numba/targets/arraymath.py#L1043 ?

rjenc29 replied:
Added tests - this should now be covered - will keep an eye out

stuartarchibald commented on numba/tests/test_array_reductions.py:
```diff
@@ -435,6 +438,112 @@ def test_nanmedian_basic(self):
+            for val in _val_variations():
+                _check_fill_diagonal(arr, val)
+
+    def _fill_diagonal_handle_unsafe_type_coercion(self):
```
This is great :)

stuartarchibald commented on numba/targets/arraymath.py:
```diff
@@ -914,6 +914,142 @@ def nanmedian_impl(arry):
+
+    # check finite values are within bounds
+    if np.any(~np.isfinite(val)) or np.any(val < v_min) or np.any(val > v_max):
+        raise ValueError('unable to safely conform val to a.dtype')
```
s/^u/U/

rjenc29 replied:
Fixed

stuartarchibald commented on numba/targets/arraymath.py:
```diff
@@ -914,6 +914,142 @@ def nanmedian_impl(arry):
+    # check finite values are within bounds
+    finite_vals = val[np.isfinite(val)]
+    if np.any(finite_vals < v_min) or np.any(finite_vals > v_max):
+        raise ValueError('unable to safely conform val to a.dtype')
```
s/^u/U/

rjenc29 replied:
Fixed

stuartarchibald commented on numba/targets/arraymath.py:
```diff
@@ -914,6 +914,142 @@ def nanmedian_impl(arry):
+def np_fill_diagonal(a, val, wrap=False):
+
+    def _abort_mission(*args):
+        raise ValueError("array must be at least 2-d")
```
s/^a/A/

rjenc29 replied:
Removed altogether - letting this fail at compile time

stuartarchibald commented on numba/targets/arraymath.py:
```diff
@@ -914,6 +914,142 @@ def nanmedian_impl(arry):
+    def _abort_mission(*args):
+        raise ValueError("array must be at least 2-d")
+
+    def fill_diagonal_impl_scalar_val(a, val, wrap=False):
```
You may like to take a look at https://github.com/stuartarchibald/numba/commit/33b97b4ea41ff80cadf3f30998f59950af8203c0 as a way to avoid this def+branch heavy pattern? Let me know what you think, you are welcome to use/not use it.

rjenc29 replied:
Beautiful!

stuartarchibald commented:
Thanks for your status update.

1. Great, well done.
2. Thanks for trying to match something that's strange, I think what you have implemented is more than sufficient and I think NumPy needs to perhaps become a bit more consistent before more time is spent on this.
3. If you have time to follow this up with NumPy core devs given the knowledge gained from trying to replicate the function I'm sure they'd be appreciative.
4. Great, thanks.
5. hmmm, I'll have a chat with @sklam, we've seen a number of similar issues during object disposal. If we can get a reproducer to hook into it and debug that will help :)

stuartarchibald commented on numba/targets/arraymath.py:
```diff
@@ -914,6 +914,142 @@ def nanmedian_impl(arry):
+@overload(np.fill_diagonal)
+def np_fill_diagonal(a, val, wrap=False):
+
+    def _abort_mission(*args):
```
Just a thought... the dimension of the array can be picked up at compile time and Numba can simply refuse to compile it? IIRC this is the behaviour in other parts of the code base, essentially, if you can catch it at compile time, then do so.

rjenc29 replied:
Yeah, I only added this to specifically replicate the exception message you get in numpy when you give it a 1D array.  But it's a simpler and more consistent to fail at compile time, so have tweaked it to do so instead.

rjenc29 commented:
Many thanks for the detailed feeback - I will work through these items as soon as possible and revert.

rjenc29 commented on numba/tests/test_array_reductions.py:
```diff
@@ -435,6 +438,112 @@ def test_nanmedian_basic(self):
         pyfunc = array_nanmedian_global
         self.check_median_basic(pyfunc, self._array_variations)
 
+    def test_fill_diagonal(self):
```
Agreed

rjenc29 commented:
I think I've covered the review comments - just letting CI do it's thing.

I commented over in NumPy-land regarding the description of behaviour where val is an array-like and will follow this up as a background task.

Cheers!

stuartarchibald commented on numba/targets/arraymath.py:
```diff
@@ -914,6 +914,111 @@ def nanmedian_impl(arry):
+@overload(np.fill_diagonal)
+def np_fill_diagonal(a, val, wrap=False):
+
+    if a.ndim > 1:
```
Please could the `else` for this branch be added to raise a `TypingError` explicitly stating that the first argument must be at least 2-D, thanks.

Without this, the default message will appear:
```
Invalid usage of Function(<function fill_diagonal at 0x7fc72c7cd8c8>) with parameters (array(float64, 1d, C), int64)
 * parameterized
In definition 0:
    All templates rejected
[1] During: resolving callee type: Function(<function fill_diagonal at 0x7fc72c7cd8c8>)
[2] During: typing of call at test_3145.py (6)


File "test_3145.py", line 6:
def foo(x):
    np.fill_diagonal(x, 7)
```
which is fine, but we can say something specific so we probably should.

rjenc29 replied:
Oh, I see what you mean, I think - so the message now contains: 

```
TypeError: Array must be at least 2-d
```
For consistency with NumPy, I could update this to:

```
ValueError: array must be at least 2-d
```
....if preffered.

stuartarchibald replied:
Thanks for considering this change. The challenge here is to provide feedback whilst in the typing region of an `@overload`. In this part of the `@overload` only the types of the arguments are known and tracing back their actual names is prohibitively difficult from this location. As a result something like:
```python
    else:
        msg = "The first argument must be at least 2-D (found %s-D)" % a.ndim
        raise TypingError(msg)
```
tells the user that it is the first argument that is the problem (as `val` could also be an array?), explains what the problem is and what was presented. In the full context of a traceback it'll look something like:
```python
Invalid usage of Function(<function fill_diagonal at 0x7f15b04e08c8>) with parameters (array(float64, 1d, C), int64)
 * parameterized
In definition 0:
    TypingError: The first argument must be at least 2-D (found 1-D)
    raised from <path>/numba/numba/targets/arraymath.py:1023
[1] During: resolving callee type: Function(<function fill_diagonal at 0x7f15b04e08c8>)
[2] During: typing of call at test_3145.py (6)


File "test_3145.py", line 6:
def foo(x):
    np.fill_diagonal(x, 7)
```
what do you think?

rjenc29 replied:
Yep, that makes a lot of sense

stuartarchibald commented on numba/tests/test_array_manipulation.py:
```diff
@@ -585,6 +589,132 @@ def test_bad_float_index_npm(self):
 
+    def test_fill_diagonal(self):
+        self._fill_diagonal_basic()
+        self._fill_diagonal_exception_cases()
```
Perhaps move the exception cases out to another test so the disabled leak checker only applies to that test?

rjenc29 replied:
Sure, makes sense.

stuartarchibald commented:
@rjenc29 thanks for your persistence with this, code looks good, PR can be merged once CI passes. Thanks again for your contribution.
